tools/pythonpkg/src/numpy/array_wrapper.cpp

```diff
@@ -274,19 +274,37 @@ struct UUIDConvert {
 };
 
 struct ListConvert {
-	static py::list ConvertValue(Vector &input, idx_t chunk_offset, const ClientProperties &client_properties) {
-		auto val = input.GetValue(chunk_offset);
-		auto &list_children = ListValue::GetChildren(val);
-		py::list list;
-		for (auto &list_elem : list_children) {
-			list.append(PythonObject::FromValue(list_elem, ListType::GetChildType(input.GetType()), client_properties));
-		}
-		return list;
+	static py::object ConvertValue(Vector &input, idx_t chunk_offset, NumpyAppendData &append_data) {
+		auto &client_properties = append_data.client_properties;
+		auto &list_data = append_data.idata;
+
+		// Get the list entry information from the parent
+		const auto list_sel = *list_data.sel;
+		const auto list_entries = UnifiedVectorFormat::GetData<list_entry_t>(list_data);
+		auto list_index = list_sel.get_index(chunk_offset);
+		auto list_entry = list_entries[list_index];
+
+		auto list_size = list_entry.length;
+		auto list_offset = list_entry.offset;
+
+		// Initialize the array we'll append the list data to
+		auto &child_type = ListType::GetChildType(input.GetType());
+		ArrayWrapper result(child_type, client_properties, append_data.pandas);
+		result.Initialize(list_size);
+
+		// Get the child vector and append the relevant entries to our result array
+		auto child_size = ListVector::GetListSize(input);
+		auto &child_vector = ListVector::GetEntry(input);
+		D_ASSERT(list_offset + list_size <= child_size);
+		result.Append(0, child_vector, child_size, list_offset, list_size);
+		return result.ToArray();
 	}
 };
 
 struct StructConvert {
-	static py::dict ConvertValue(Vector &input, idx_t chunk_offset, const ClientProperties &client_properties) {
+	static py::dict ConvertValue(Vector &input, idx_t chunk_offset, NumpyAppendData &append_data) {
+		auto &client_properties = append_data.client_properties;
```

Comment:
We are losing the source offset in `StructConvert`, `MapConvert`, `UnionConvert`, etc. Will this work correctly when these types are present in lists? Can we add some tests?

Reply:
I think we are adding it to the `chunk_offset` passed to `ConvertNested` which calls this, but I'll check never the less 👍